src/java_bytecode/generate_java_generic_type.cpp

```diff
@@ -189,40 +207,129 @@ irep_idt generate_java_generic_typet::build_generic_tag(
   const java_class_typet &original_class) const
 {
   std::ostringstream new_tag_buffer;
-  new_tag_buffer << original_class.get_tag();
-  new_tag_buffer << "<";
-  bool first=true;
-  for(const typet &type_argument : existing_generic_type
-    .generic_type_arguments())
+  const std::string &original_class_tag = original_class.get_tag().c_str();
+  auto generic_argument_p =
+    existing_generic_type.generic_type_arguments().begin();
+
+  // if the original class is implicitly generic, add type
+  // arguments for all generic outer classes
+  if(is_java_implicitly_generic_class_type(original_class))
   {
-    if(!first)
-      new_tag_buffer << ", ";
-    first=false;
-
-    INVARIANT(
-      !is_java_generic_parameter(type_argument),
-      "Only create full concretized generic types");
-    const irep_idt &id(id2string(type_argument.subtype().get(ID_identifier)));
-    new_tag_buffer << pretty_print_java_type(id2string(id));
-    if(is_java_array_tag(id))
+    const java_implicitly_generic_class_typet
+      &implicitly_generic_original_class =
+        to_java_implicitly_generic_class_type(original_class);
+    const auto &implicit_generic_types =
+      implicitly_generic_original_class.implicit_generic_types();
```

Comment:
It might be worth documenting here that you are assuming that the order of implicit generic types is outer to inner (I think this is required for your for loop to work - you could alternatively make the for loop somehow build up each classes generic tag in parallel allowing you to jump around.

Reply:
Yes, that is true. At the moment, the implicit generic types are ordered, the code for the creation of this vector was part of the previous PR https://github.com/diffblue/cbmc/pull/1616. There is a comment there that the types are ordered. Would making a comment here as well be sufficient then?

Reply:
Yup